ports/esp32/main/CMakeLists.txt

```diff
@@ -16,7 +16,7 @@ if(NOT CMAKE_BUILD_EARLY_EXPANSION)
 endif()
 
 set(MICROPY_QSTRDEFS_PORT
-    ${PROJECT_DIR}/qstrdefsport.h
+    ${CMAKE_CURRENT_LIST_DIR}/../qstrdefsport.h
```

Comment:
Maybe this can be `${MICROPY_PORT_DIR}/qstrdefsport.h`, because that's exactly where it's supposed to look, in the *port* directory.

Reply:
I'm glad you mentioned this.  On line 3 there is currently a check for `MICROPY_DIR` and setting the value if unset.  If that had been the case and `MICROPY_DIR` had *not* been set, how would `MICROPY_PORT_DIR` be set?

Reply:
I assume something like this right below line 3 would be OK?

```Python3
if(NOT MICROPY_PORT_DIR)
   get_filename_component(MICROPY_PORT_DIR ${CMAKE_CURRENT_LIST_DIR}/.. ABSOLUTE)
endif()
```

Reply:
But.. `${MICROPY_PORT_DIR}` is set in `ports/esp32/CMakeLists.txt`, so shouldn't the value of `${MICROPY_PORT_DIR}` be available in this file?

> I ask additionally because this is a different directory with different scope from the parent directory.

Does CMake scope variables per-CMakeLists.txt file?  If that's so, how do `extmod.cmake`, `mkrules.cmake` have access to this variable?

Reply:
TL;DR: The ESP-IDF messes things up a bit.  They make the "entrypoint" for consumption _slightly_ different from RP2, Zephyr, et al.

The choice appears to be 
1) We _solely_ use `MICROPY_PORT_DIR` and force people to RTFM and define `MICROPY_PORT_DIR` themselves on ESP32 (Note: In testing this lead to cryptic errors)
2) We see if there is some way to calculate the path for all ports.
3) We move the following proposed block from the PR into the "component", leaving the original definition in the project.  This allows for my original intention of providing a re-calculation of the path if it's not been populated.

```cmake
if(NOT MICROPY_PORT_DIR)
    set(MICROPY_PORT_DIR ${CMAKE_CURRENT_LIST_DIR}/..)
endif()
```

4) We split the difference between 1 & 3 and instead emit a warning letting users know that they have not defined a required value:

```cmake
if(NOT MICROPY_PORT_DIR)
    message(WARNING "Variable MICROPY_PORT_DIR has not been set, some paths may not enumerate as expected")
endif()
```

---
In depth:

CMake has a number of variable scopes, a specific "Directory" scope is one of the most common but your question about `extmod.cmake` and `mkrules.cmake` bring up an important distinction.

In the case of `extmod.cmake` and `mkrules.cmake` they're considered [Modules](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#id13):

> CMake Language code in either [Directories](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#directories) or [Scripts](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#scripts) may use the [`include()`](https://cmake.org/cmake/help/latest/command/include.html#command:include "include") command to load a `<module>.cmake` source file in the scope of the including context. See the [`cmake-modules(7)`](https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#manual:cmake-modules(7) "cmake-modules(7)") manual page for documentation of modules included with the CMake distribution. Project source trees may also provide their own modules and specify their location(s) in the [`CMAKE_MODULE_PATH`](https://cmake.org/cmake/help/latest/variable/CMAKE_MODULE_PATH.html#variable:CMAKE_MODULE_PATH "CMAKE_MODULE_PATH") variable.

For ESP32 projects external to MicroPython, the point of consumption will be `ports/esp32/main/CMakeLists.txt` as opposed to `ports/esp32/CMakeLists.txt` due to ~~hijacking~~ opinionation by the ESP-IDF.  

Users will be consuming the "component" (the `main` subdirectory) and not the "project" (what we think of as `MICROPY_PORT_DIR`).  That means that the directory scope starts at `main`.  As to the modules, since they are pulled in *explicitly* via calls to `include()` their scope is maintained.

I realized that in my test "project" ([mp-test](https://github.com/brianredbeard/mp-test/blob/05606c2fcf6f53f12fcdf18cc214fa65fcfff34b/CMakeLists.txt#L33)) that I had taken the liberty of defining `MICROPY_PORT_DIR` which took care of ensuring this value would exist when consumed in the component.  Simultaneously, the CI builds succeeded because `MICROPY_PORT_DIR` was defined in the native "project".  

I tested this by removing the `set(MICROPY_PORT_DIR` call in my `mp-test` _project_ and replacing instances of `${CMAKE_CURRENT_LIST_DIR}/../` with `${MICROPY_PORT_DIR}` in the _component_ (but not defining the value in the component as proposed previously - e.g.  ).

After this change when attempting to configure the build it fails with the following error:

```cmake
-- Configuring done (4.9s)
CMake Error at /home/bharrington/Projects/mb-badgepython/esp-idf/tools/cmake/component.cmake:478 (add_library):
  Cannot find source file:

    //main.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
Call Stack (most recent call first):
  external/micropython/ports/esp32/main/CMakeLists.txt:159 (idf_component_register)
  components/micropython/CMakeLists.txt:10 (include)


-- Generating done (0.5s)
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

To troubleshoot this, I can generate a full trace of the build steps with diagnostic info:

`cmake -S . -B build --trace-expand --trace-redirect=build.out -G Ninja`

This generates about 66MB of output, and buried within that we see:

 `584853:/home/bharrington/Projects/mp-test/external/micropython/ports/esp32/main/CMakeLists.txt(51):  set(MICROPY_SOURCE_PORT /main.c /uart.c /usb.c /usb_serial_jtag.c /gccollect.c /mphalport.c /fatfs_port.c /help.c /machine_bitstream.c /machine_timer.c /ma
chine_pin.c /machine_touchpad.c /machine_adc.c /machine_adcblock.c /machine_dac.c /machine_i2c.c /machine_i2s.c /machine_uart.c /modmachine.c /network_common.c /network_lan.c /network_ppp.c /network_wlan.c /mpnimbleport.c /modsocket.c /modesp.c /esp32_nvs.
c /esp32_partition.c /esp32_rmt.c /esp32_ulp.c /modesp32.c /machine_hw_spi.c /machine_wdt.c /mpthreadport.c /machine_rtc.c /machine_sdcard.c /modespnow.c )`

So in this case, ***not*** ensuring `MICROPY_PORT_DIR` is set at the correct scope adds additional challenge to troubleshooting because the last hint of where something went wrong was at `idf_component_register`, over 100 lines after the faulty definition of `MICROPY_SOURCE_PORT`.

It was running into errors like that which sent me down the CMake rabbit hole.  🤔 😀

Reply:
Thanks for the detailed investigation.  I've been down a CMake rabbit hole many times before...

I think option 3 is the best (put the definition of `MICROPY_PORT_DIR` in the component in `ports/esp32/main/CMakeLists.txt`).

But also I think we can just remove the definition of this variable from the project (from `ports/esp32/CMakeLists.txt`), and where it was previously used in that file (there are two locations) write instead `CMAKE_SOURCE_DIR`.  These two uses are setting the default board dir and default manifest, which are already overridable by the user.

(My aim here is to keep things as simple as possible, as far as that is possible with CMake.)